test/solver/batch_bicgstab_kernels.cpp

```diff
@@ -0,0 +1,232 @@
+                                                      num_rhs, 0, ref_iters);
+
+    auto res = gko::test::solve_linear_system(exec, solve_lambda,
+                                              solver_settings, linear_system);
+
+    auto iter_array = res.log_data->iter_counts.get_const_data();
+    for (size_t i = 0; i < num_batch_items; i++) {
+        ASSERT_EQ(iter_array[i], ref_iters);
+    }
+}
+
+
+TEST_F(BatchBicgstab, CanSolve3ptStencilSystem)
+{
+    const int num_batch_items = 12;
+    const int num_rows = 100;
+    const int num_rhs = 1;
+    const real_type tol = 1e-5;
+    const int max_iters = 100;
+    auto linear_system = setup_linsys_and_solver<Mtx>(num_batch_items, num_rows,
+                                                      num_rhs, tol, max_iters);
+    auto solver = gko::share(solver_factory->generate(linear_system.matrix));
+
+    auto res = gko::test::solve_linear_system(exec, linear_system, solver);
+
+    GKO_ASSERT_BATCH_MTX_NEAR(res.x, linear_system.exact_sol, tol * 10);
+    for (size_t i = 0; i < num_batch_items; i++) {
+        auto comp_res_norm =
+            exec->copy_val_to_host(res.res_norm->get_const_values() + i) /
+            exec->copy_val_to_host(linear_system.rhs_norm->get_const_values() +
+                                   i);
+        ASSERT_LE(comp_res_norm, tol);
+    }
+}
+
+
+TEST_F(BatchBicgstab, CanSolveLargeHpdSystem)
+{
+    const int num_batch_items = 3;
+    const int num_rows = 1025;
+    const int num_rhs = 1;
+    const real_type tol = 1e-5;
+    const int max_iters = 2000;
+    const real_type comp_tol = tol * 100;
+    auto solver_factory =
+        solver_type::build()
+            .with_default_max_iterations(max_iters)
+            .with_default_tolerance(tol)
+            .with_tolerance_type(gko::batch::stop::ToleranceType::absolute)
+            .on(exec);
+    std::shared_ptr<Logger> logger = Logger::create(exec);
+    auto linear_system = gko::test::generate_diag_dominant_batch_problem<Mtx>(
+        exec, num_batch_items, num_rows, num_rhs, true);
+    auto solver = gko::share(solver_factory->generate(linear_system.matrix));
+    solver->add_logger(logger);
+
+    auto res = gko::test::solve_linear_system(exec, linear_system, solver);
+
+    solver->remove_logger(logger);
+    auto iter_counts = gko::make_temporary_clone(exec->get_master(),
+                                                 &logger->get_num_iterations());
+    auto res_norm = gko::make_temporary_clone(exec->get_master(),
+                                              &logger->get_residual_norm());
+    GKO_ASSERT_BATCH_MTX_NEAR(res.x, linear_system.exact_sol, comp_tol);
+    for (size_t i = 0; i < num_batch_items; i++) {
+        auto comp_res_norm =
+            exec->copy_val_to_host(res.res_norm->get_const_values() + i);
+        ASSERT_LE(iter_counts->get_const_data()[i], max_iters);
+        EXPECT_LE(res_norm->get_const_data()[i], comp_tol);
```

Comment:
why does this criterion need use `100 * tol` not tol if the criterion is absolute residual norm?

Reply:
I think there were issues only on some systems, particularly MSVC. Not sure why.

Reply:
It's might related to the optimization or different random input?
The codes gives me the confusion about the criterion.
From my first thought, it is actual residual norm check. That's why I do not think that the residual norm does not match the required criterion makes sense.

Reply:
I think this code is a bit stale and has been updated. So, I think it should be correct now. In the updated code, `comp_res_norm` is the actual residual while `resnorm` is the residual from the logger, which in this case is  the implicit residual.